### PR TITLE
Account Settings

### LINKED ISSUE #3726: Account Settings
We'll be looking to improve the Settings page to add a couple more features, update it's style and refactor out some old code. The work is already ongoing on `feature/account-settings` branch, and it's a bit late to open the issue. Here is a screenshot to show some of the changes we're looking to implement:

android-account-settings

### PR DESCRIPTION
**Pull request by oguzkocer**
Fixes #3726. This PR is part of the Calypsofication process of `/me`. It adds a couple new fields to Settings page such as username, email, primary site & web address. It also refactors language handling to match Site Settings and reuse existing components. I found 2 things I wasn't sure how to handle, so I saved them to the PR review.

1. In the original code, when a language other than the device's locale is selected, we track an event in the Analytics. I've moved this code to the new implementation. However I have some concerns about it. If the user changes the language to non-default and then changes it back to the default language, we don't keep track of it. I don't know if this is important or not, but I thought I'd mention it. /cc @daniloercoli & @tonyr59h
2. I wanted to change the language and the primary site after the user closed the dialog, so I've added an event for it (`ListPreferenceDialogClosed`). After this implementation, I thought the interface pattern might have made more sense. But before changing it, I wanted to get a second opinion.

Here is a screenshot of the current design: https://cloudup.com/cCxJ3ab8G4P /cc @rickybanister & @mattmiklic 

Note: I've split this PR to a couple PRs and it's still a bit bigger than I'd like. If it's messy to review, please let me know and I'll do my best to split a few more things.

/cc @astralbodies

### CODE REVIEW

**Comment by daniloercoli** [comment]
>  If the user changes the language to non-default and then changes it back to the default language, we don't keep track of it. I don't know if this is important or not, but I thought I'd mention it. 

I think we should track it.

**Comment by oguzkocer** [comment]
> I think we should track it.

I opened a new issue #3768 for this and will take care of it once this PR is merged.

File: WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java [hunk 1]
```diff
@@ -177,12 +178,16 @@ public static void addMedia(Activity activity) {
...
     public static void viewMyProfile(Context context) {
+        AccountHelper.getDefaultAccount().fetchAccountSettings();
+
         Intent intent = new Intent(context, MyProfileActivity.class);
         AnalyticsUtils.trackWithCurrentBlogDetails(AnalyticsTracker.Stat.OPENED_MY_PROFILE);
         slideInFromRight(context, intent);
     }
 
     public static void viewAccountSettings(Activity activity) {
+        AccountHelper.getDefaultAccount().fetchAccountSettings();
```
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java]: Hey @oguzkocer, these Account settings update calls feel strange be initiated here. I think the helper methods here in `ActivityLauncher.java` should only launch the activities without too much business logic. May I suggest moving the update calls either closer to the caller or in the Activities themselves? Cheers!

Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java]: I kind of disagree with that. Moving to the activities will mean that if the activity is restarted, we'll be fetching the details again and keeping the fetch here prevents that. I'd also argue that, it's not that different than adding the analytics call here. I'd consider this as more of a central dispatcher than a helper class which just creates the intents.

In an ideal case, the lower layers would decide when to do fetching with the rules we specify (unless it's forced by the user). However, with the current network structure, that's not easy to do.

I can still move it to the Activity, if you feel strongly about it, but I thought I'd at least share my point of view. Cheers!

Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java]: I would actually prefer my first suggestion, which is to be closer to the caller.

Here's the thing: `viewAccountSettings` seems innocent enough (it has "view" in its name) and is in an activity launcher helper class. But, it also changes the local world since it fetches and updates local data. That's quite different than spawning an analytics event and it couples the activity opening with a network action. I would feel way more comfortable if those two operations were not coupled like that :)

Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java]: The caller in this case is `MeFragment` and I don't think it makes a lot of sense to start a fetch there. I also think it defeats the whole purpose of `ActivityLauncher` which is to remove duplication and make sure the activities are launched in a consistent way across the app. If we were to add "My Profile" into the signup process, it'd mean that we'd also need to start a fetch from there.

The best way to solve this would be to let the lower layers handle the regular fetch only when it's necessary. But out of 3 different places where we can put this code right now, I think this one makes the most sense. I could be totally wrong though :)

@maxme any thoughts on this?

Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java]: It's seems that this is more like a matter of preference and judging by the low usage count of these functions in the current codebase/PR I think it doesn't warrant blocking the PR on it.

**Comment by maxme** [comment]
IIRC, `ActivityLauncher` was introduced to keep the launch animations consistent for each activity (it also helps by killing some duplicate code). I don't have strong opinion on this point, it's a detail, choose your preferred solution ;). Are you listening for any events emitted by `fetchAccountSettings` in the activity, if yes, it makes sense to launch the call from the activity itself. 

Sidenote: We should also skip the call there if the network is down.

File: WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java (hunk 1 above)
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java]: IIRC, `ActivityLauncher` was introduced to keep the launch animations consistent for each activity (it also helps by killing some duplicate code). I don't have strong opinion on this point, it's a detail, choose your preferred solution ;). Are you listening for any events emitted by `fetchAccountSettings` in the activity, if yes, it makes sense to launch the call from the activity itself. 

Sidenote: We should also skip the call there if the network is down.

Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java]: Thanks for the input @maxme & @hypest! I moved it to `onStart` in [7cb9610](https://github.com/wordpress-mobile/WordPress-Android/commit/7cb961069f881a6709edec0fcd7936373de70615). I hope that's the correct place!

**Comment by maxme** [comment]
Can we rename `SettingsActivity` to `AccountSettingsActivity`, I'm lost in our (site/notification/account) settings ;)

**Comment by oguzkocer** [comment]
@maxme: I renamed both `SettingsActivity` & `SettingsFragment` to their Account Settings counterpart. Thanks!

**Comment by tonyr59h** [comment]
Looking good! Very similar feel to the Site Settings list, I like the consistency :smile:

Found a few functional issues and I've got some questions on the design that I'll make a separate post for.

* Updating the email address or primary site clears the web address field. Leaving the settings screen then returning will show the correct value
* I've run into a weird situation with the email address field. Account starts with `tonyr@tonyrankin.com`. Set it to `tony.rankin@a8c.com` on the web. Refresh app settings page (leave, join), app still shows `tonyr@tonyrankin.com`. Attempt to change to `tony.rankin@a8c.com`, field does not disable. Attempt to change to `tonyr@tonyrankin.com`, field does disable. [Video](https://cldup.com/aFcgE6o05A.mp4)
* The above video also highlights an `EditText` issue, we should change its input type to `textEmailAddress`
* I suspect the second bullet is caused by the way `AccountModel` updates its email address [here](https://github.com/wordpress-mobile/WordPress-Android/blob/feature/account-settings/WordPress/src/main/java/org/wordpress/android/models/AccountModel.java#L80).
* Selecting `Revert` after changing the email address makes the primary site dialog have no selection when opened
* Selecting a new value in the primary site dialog does not deselect the previous option box

**Comment by tonyr59h** [comment]
Design questions cc @mattmiklic

* Action Bar font
* Preference title capitalization choices. I think they should be consistent with the web, so `Email address` would be `Email Address`, etc...
* We should use single line `EditText`'s for email address and web address dialogs
* Should the primary site show the title of the blog instead of the URL?
* Dialog title fonts should match, email address and web address should use custom titles?

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/EditTextPreferenceWithValidation.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+                    if (error != null) {
+                        getEditText().setError(error);
+                    } else {
+                        callChangeListener(text);
+                        dialog.dismiss();
+                    }
+                }
+            });
+            positiveButton.setTextColor(buttonColor);
```
Comment (tonyr59h) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/EditTextPreferenceWithValidation.java]: There are utility methods you can use for these buttons. [Example here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/prefs/SummaryEditTextPreference.java#L130).

**Comment by mattmiklic** [comment]
> Action Bar font
It should be Open Sans Semibold in the same size and color it is now, though we may be on the verge of switching to system fonts so it's not a big concern for me.

> Preference title capitalization choices. I think they should be consistent with the web, so Email address would be Email Address, etc...
Should definitely be consistent; I talked with @drw158 at one point about sentence case vs. title case for labels, and I think he made a good argument at the time (though I don't remember it exactly at the moment) for using sentence case. Right now, I'd probably just go with whatever is most common elsewhere in the app, for the sake of consistency.

Will let @rickybanister answer the other questions since he designed this section and is more familiar with it.

**Comment by hypest** [comment]
Lemme add another design side question for @rickybanister, regarding the "WordPress for Android" screen this time:
There is the very similar looking "Help & Support" screen, higher in the view hierarchy, which also btw displays the app version name. I even felt that we kinda have two "About" screens that way (which ofc is not really accurate). So the question is, would it make sense to reconcile those two together somehow? A quick idea would be to launch the "WordPress for Android" page from the help page instead of Account Settings.

In any case, this is not dramatic enough to hold back this PR on it. Cheers!

**Comment by hypest** [comment]
May I suggest pre-filling in the email and web address popup EditTexts with the current value of the respective setting? Cheers!
While at it, I would also suggest clearing the EditText's error string before popping up. Currently, if user has inputted an error email and dismissed the dialog, next time it pops up it still has the error displayed, which most probably is not what's intended.

**Comment by hypest** [comment]
I have most probably missed any discussion/decision on this one but that's actually a good thing in this case:
Tapped the `Turn PIN lock on` and I was presented by (an awesomely looking screen, well done @oguzkocer !) asking me: `Enter your PIN`. What PIN would that be? My device pin? I'm sure not, but it might make lots of users sit there and think about it for a sec and still not be sure what this means or does. And what exactly will that lock down after I enter a PIN?
I guess what I'm trying to say is that I could use a bit more detail here. Something like "Enter a new app PIN" (if I do it the first time around) or "Enter your current app PIN" if i'm going to change it. I'll leave to the more design+english fluent here to come up with the best wording ;)

**Comment by rickybanister** [comment]
> So the question is, would it make sense to reconcile those two together somehow? A quick idea would be to launch the "WordPress for Android" page from the help page instead of Account Settings.

I wanted us to have the About pages be here to match the iOS navigation a bit more while attempting to keep things similar to Calypso. I suppose ideally Help & Support wouldn't have an intermediary step with the About info and simply give you access to help info and contacting support. Obviously that is a larger change, so I'm not 100% sure the best approach. I almost feel that since the `WordPress for Android` section is finished and in the Account area it is ok to have it redundant for a short time. 

> We should use single line EditText's for email address and web address dialogs

I'm not 100% sure of the pattern, but there is no need to have multiline inputs for single line text, so yes I think.

> Should the primary site show the title of the blog instead of the URL?

It would be alright to show the site title in the list view—ideally in the selection dialog we could show both title and url though.

**Comment by oguzkocer** [comment]
@hypest: I've tried changing the input type in several places and it doesn't look like it's working. Any ideas why that might be?

Note: For the multi-line question, I am assuming if we can set the web address's input type, it should prevent the 2nd line, if it doesn't I'll handle it.

**Comment by oguzkocer** [comment]
> Should the primary site show the title of the blog instead of the URL?
> > It would be alright to show the site title in the list view—ideally in the selection dialog we could show both title and url though.

@rickybanister: The selection dialog already shows both the title and the address. Should I change the one under the "Primary Site" to the title instead of the url?

**Comment by oguzkocer** [comment]
@hypest & @rickybanister: Related to the discussion about the placement of About section, I'd suggest to open a new issue and discuss it under it to avoid noise in the PR. Whichever way we decide to go, I'd rather not handle it inside the PR as it's already big enough. Thanks!

**Comment by oguzkocer** [comment]
> I've run into a weird situation with the email address field. Account starts with tonyr@tonyrankin.com. Set it to tony.rankin@a8c.com on the web. Refresh app settings page (leave, join), app still shows tonyr@tonyrankin.com. Attempt to change to tony.rankin@a8c.com, field does not disable. Attempt to change to tonyr@tonyrankin.com, field does disable.

@tonyr59h You should see a Snackbar once you change your email from either web or the app, the email field will stay the same until it's verified but in the Snackbar we tell the user that there is a change pending. Here is a screenshot: https://cloudup.com/cCJgVmfIcXy

**Comment by oguzkocer** [comment]
> Preference title capitalization choices. I think they should be consistent with the web, so Email address would be Email Address, etc...

@tonyr59h I initially used "First Name" in My Profile and after discussion it was changed to "First name", so I kept it consistent with it in Account Settings. If we think capitalized version makes more sense, we can open a new issue and change both My Profile & Account Settings together. Does that work for you?

**Comment by oguzkocer** [comment]
@tonyr59h I think I've either fixed or replied to all of the feedback. I've one last question about the actionbar title. The titles look the same font & size to me in between Site Settings - Account Settings. Are you sure there is an issue there? (I don't really trust myself) Should I just use `site_settings_actionbar.xml` for the layout of the actionbar? (and maybe rename it to `preference_actionbar.xml`)

Basically do this: https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/prefs/BlogPreferencesActivity.java#L71-L76

**Comment by tonyr59h** [comment]
> he made a good argument at the time (though I don't remember it exactly at the moment) for using sentence case

The [design guidelines](https://www.google.com/design/spec/style/writing.html#writing-language) recommend using sentence case, that's probably the best argument for it.

> If we think capitalized version makes more sense, we can open a new issue and change both My Profile & Account Settings together. Does that work for you?

Based on @mattmiklic's feedback we should definitely go for consistency. The current titles work for now, we should open an issue to revisit Site Settings, Account Settings, and My Profile to convert them all the sentence case.

**Comment by tonyr59h** [comment]
Awesome job addressing all the comments and addressing some other issues in the process (`DetailListPreference` :anger:)! It looks like things are working well, :+1: for a merger once Travis does its thing.

**Comment by oguzkocer** [comment]
Thanks for the review and walking through me the last bit of issues on Slack @tonyr59h! I really appreciate it!

Thank you @rickybanister, @mattmiklic & @hypest for your input. There have been a couple discussions about the Account Settings page and if you guys think we should change/improve something, let's open a new issue and keep the discussion there. Please feel free to ping me for the implementation of those as well, I am happy to take care of them on a separate PR.

**Comment by tonyr59h** [comment]
:shipit:
